catalogue_graph/src/data_validation/concept_types.py

```diff
@@ -0,0 +1,59 @@
+from collections.abc import Generator
+
+from models.graph_node import ConceptType
+from utils.aws import get_neptune_client
+
+CONCEPT_TYPES_QUERY_LIMIT = 20000
+CONCEPT_TYPES_QUERY = """
+    MATCH (concept:Concept)
+    WITH concept ORDER BY concept.id
+    SKIP $start_offset LIMIT $limit
+    OPTIONAL MATCH (work)-[has_concept:HAS_CONCEPT]-(concept)
+    WITH 
+        concept,
+        collect(DISTINCT has_concept.referenced_type) AS concept_types
+    RETURN
+        concept,
+        concept_types        
+"""
+
+
+def _are_concept_types_consistent(concept_types: list[ConceptType]) -> bool:
+    """Return `True` if all provided concept types are mutually compatible. Otherwise, return `False`."""
+    # 'Concept' and 'Subject' types are consistent with all other types,
+    # so we filter them out when determining consistency
+    filtered_types = [c for c in concept_types if c not in ("Concept", "Subject")]
+
+    if len(filtered_types) <= 1:
+        return True
+
+    # Of all remaining types, only Agent/Org and Agent/Person are compatible. All other combinations are invalid.
+    compatible_combinations = [{"Agent", "Organisation"}, {"Agent", "Person"}]
+    return set(filtered_types) in compatible_combinations
+
+
+def _get_concepts_count() -> int:
+    count_query = "MATCH (c:Concept) RETURN count(c) AS count"
+    response = get_neptune_client(True).run_open_cypher_query(count_query)
+    count: int = response[0]["count"]
+    return count
+
+
+def get_concepts_with_inconsistent_types() -> Generator[dict]:
+    """Return all concepts whose combination of types is not consistent."""
+    client = get_neptune_client(True)
+
+    start_offset = 0
+    while start_offset < _get_concepts_count():
```

Comment:
Should the result of `_get_concepts_count` be stored as a variable outside the loop, or is it expected to keep updating with each iteration?  If the latter, it could do with a comment, because this is the kind of thing someone might come along and "fix"